### ISSUE TITLE
Console runner - Explore flag does not apply where filter to output

### ISSUE THREAD

**Issue by ChrisMaddock** [nunit/nunit-console#139]
@chriscore commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143)

Hi there

Running nunit3-console.exe with arguments similar to the following does not apply the filter specified in /where to the results which are seen on the output.

"C:\MyRepo\MyAssembly.dll" /explore /noheader /where="cat==Awesome"

All tests are returned by /explore, including those which do not satisfy cat==Awesome

This was found in NUnit Console Runner v3.0.5813

---

@CharliePoole commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165499679)

I was uncertain whether it made sense to use a filter when listing tests, but I suppose it does if you tried it!

---

@chriscore commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165501578)

Hey Charlie

Thanks, before a test run I am currently trying to call this to see how many tests my filter would return - it would certainly be very useful for me at this point :)

Great work with nunit 3 so far

Chris

---

@CharliePoole commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165502901)

@chriscore If you have a runner, just call CountTestCases on it with your filter!

If you are simply scripting nunit3-console, that's not an option, of course.

---

@chriscore commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165505549)

@CharliePoole I don't currently have a runner instance as I haven't been able to figure out how to get that going, like you say I have just scripted nunit3-console, but having that control in process would be ideal.

I did find it difficult to find how to host a runner as there seemed to be a lot of dependent objects which I found difficult to create, do you have a C# snippet anywhere which might give some clues how to start? Being able to load and run tests in process would be ideal for what I'm trying to do here.

---

@CharliePoole commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165509169)

Add a reference to nunit.engine.api.dll. Then (simplified)...

```
ITestEngine engine = TestEngineActivator.CreateInstance();
TestPackage package = new TestPackage(ASSEMBLY_PATH);
ITestRunner runner = engine.GetRunner(package);
```

This leaves out setting options, using a filter, etc. For the full story, start here:
    https://github.com/nunit/nunit/blob/master/src/NUnitConsole/nunit3-console/Program.cs#L111

---

@chriscore commented on [Thu Dec 17 2015](https://github.com/nunit/nunit/issues/1143#issuecomment-165511283)

Thanks so much for that, I didn't know about the TestEngineActivator!

---

@rprouse commented on [Thu Apr 07 2016](https://github.com/nunit/nunit/issues/1143#issuecomment-206954366)

When this is fixed, ensure it also works with NUnitLite. See #1213

---

@NCarter3 commented on [Tue Jun 21 2016](https://github.com/nunit/nunit/issues/1143#issuecomment-227305222)

I'd also love to see filters working with --explore!

**Comment by CharliePoole** [comment]
Note that this can't be fixed entirely in the console project. The console and engine can pass a filter to the framework, but it's up to the framework to do something with it. Issue nunit/nunit#1143 will most likely need to be fixed before this one.

**Comment by Poimen** [comment]
@ChrisMaddock  @CharliePoole  As per out brief discussion on the nunit-discussion forum, I'll be happy to have a look into this issue for you.

**Comment by ChrisMaddock** [comment]
Have assigned you to both issues @Poimen - thanks! 😄

**Comment by kdubau** [comment]
Is there any update on the progress of this feature? @Poimen

**Comment by kdubau** [comment]
Looks like https://github.com/nunit/nunit/pull/2094 is in progress as of last week.

**Comment by Poimen** [comment]
@kdubau Unfortunately the going has been very very slow due to a number of external issues that I haven't been able to focus on this like I would have liked.

At the moment the nunitlite implementation is being discussed but still battling the XML outputting of the reporting.

**Comment by Poimen** [comment]
@kdubau PR nunit/nunit#2094 has been merged which addresses this issue 😃

**Comment by ChrisMaddock** [comment]
The framework changes have been merged and released in NUnit Framework v3.7, and we decided on #208, that there wasn't much to put in a unit test in this repo. Thanks @Poimen!

**Comment by svetikshark** [comment]
trying to use --explore  with --where in console runner 3.8.0 and still it returns all test names
/d/src/ja-automation/SeleniumTest/packages/NUnit.ConsoleRunner.3.8.0/tools/tools 
$ nunit3-console.exe "d:\src\ja-automation\SeleniumTest\SeleniumTest\bin\Debug\
SeleniumTest.dll" --where 'cat == Smoke'   --explore:results

results 
<test-run id="2" testcasecount="2399">
